daprovider/server/provider_server.go

```diff
@@ -0,0 +1,193 @@
+	f.Bool(prefix+".enable-da-writer", DefaultServerConfig.EnableDAWriter, "implies if the das server supports daprovider's writer interface")
+	f.Int(prefix+".rpc-server-body-limit", DefaultServerConfig.RPCServerBodyLimit, "HTTP-RPC server maximum request body size in bytes; the default (0) uses geth's 5MB limit")
+	genericconf.HTTPServerTimeoutConfigAddOptions(prefix+".server-timeouts", f)
+}
+
+func fetchJWTSecret(fileName string) ([]byte, error) {
+	if data, err := os.ReadFile(fileName); err == nil {
+		jwtSecret := common.FromHex(strings.TrimSpace(string(data)))
+		if len(jwtSecret) == 32 {
+			log.Info("Loaded JWT secret file", "path", fileName, "crc32", fmt.Sprintf("%#x", crc32.ChecksumIEEE(jwtSecret)))
+			return jwtSecret, nil
+		}
+		log.Error("Invalid JWT secret", "path", fileName, "length", len(jwtSecret))
+		return nil, errors.New("invalid JWT secret")
+	}
+	return nil, errors.New("JWT secret file not found")
+}
+
+// NewServerWithDAPProvider creates a new server with pre-created reader/writer/validator components
+func NewServerWithDAPProvider(ctx context.Context, config *ServerConfig, reader daprovider.Reader, writer daprovider.Writer, validator daprovider.Validator) (*http.Server, error) {
+	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", config.Addr, config.Port))
+	if err != nil {
+		return nil, err
+	}
+
+	rpcServer := rpc.NewServer()
+	if config.RPCServerBodyLimit > 0 {
+		rpcServer.SetHTTPBodyLimit(config.RPCServerBodyLimit)
+	}
+
+	server := &Server{
+		reader:    reader,
+		writer:    writer,
+		validator: validator,
+	}
+	if err = rpcServer.RegisterName("daprovider", server); err != nil {
+		return nil, err
+	}
+
+	addr, ok := listener.Addr().(*net.TCPAddr)
+	if !ok {
+		return nil, errors.New("failed getting provider server address from listener")
+	}
+
+	var handler http.Handler
+	if config.JWTSecret != "" {
+		jwt, err := fetchJWTSecret(config.JWTSecret)
+		if err != nil {
+			return nil, fmt.Errorf("failed creating new provider server: %w", err)
+		}
+		handler = node.NewHTTPHandlerStack(rpcServer, nil, nil, jwt)
+	} else {
+		handler = rpcServer
+	}
+
+	srv := &http.Server{
+		Addr:              "http://" + addr.String(),
+		Handler:           handler,
+		ReadTimeout:       config.ServerTimeouts.ReadTimeout,
+		ReadHeaderTimeout: config.ServerTimeouts.ReadHeaderTimeout,
+		WriteTimeout:      config.ServerTimeouts.WriteTimeout,
+		IdleTimeout:       config.ServerTimeouts.IdleTimeout,
+	}
+	go func() {
+		if err := srv.Serve(listener); err != nil &&
+			!errors.Is(err, http.ErrServerClosed) {
+			log.Error("provider server's Serve method returned a non http.ErrServerClosed error", "err", err)
+		}
+	}()
+
+	go func() {
+		<-ctx.Done()
+		_ = srv.Shutdown(context.Background())
+	}()
+
+	return srv, nil
+}
+
+func (s *Server) IsValidHeaderByte(ctx context.Context, headerByte byte) (*daclient.IsValidHeaderByteResult, error) {
```

Comment:
not an absolute must, but I don't like the server importing the client library for request/respond types. It creates a weird dependency IMHO.
I prefer the relevant types sit in their own separate library for just the common interface.
See for example: validator/server_api/json.go

Reply:
Good idea, done.